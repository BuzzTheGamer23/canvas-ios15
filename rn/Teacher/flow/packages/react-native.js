// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
