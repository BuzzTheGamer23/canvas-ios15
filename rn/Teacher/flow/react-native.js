//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
