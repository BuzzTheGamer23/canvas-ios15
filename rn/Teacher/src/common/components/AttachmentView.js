//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/**
 * @flow
 */

import React, { Component } from 'react'
import {
  View,
  WebView,
  Image,
  StyleSheet,
  Text,
  ActionSheetIOS,
  ActivityIndicator,
} from 'react-native'
import {
  downloadFile,
  CachesDirectoryPath,
  stopDownload,
} from 'react-native-fs'
import i18n from 'format-message'

import Screen from '../../routing/Screen'
import Colors from '../colors'
import Images from '../../images'
import Navigator from '../../routing/Navigator'
import Video from './Video'

type Props = {
  attachment: Attachment,
  style?: any,
  navigator: Navigator,
}

export default class AttachmentView extends Component<any, Props, any> {
  state = {
    size: { width: 0, height: 0 },
    filePath: null,
  }

  componentDidMount () {
    if (this.props.attachment.uri) {
      this.setState({ filePath: this.props.attachment.uri })
      return
    }
    const path = `${CachesDirectoryPath}/${this.props.attachment.filename || this.props.attachment.display_name}`
    let { jobId, promise } = downloadFile({
      fromUrl: this.props.attachment.url,
      toFile: path,
    })
    this.setState({ jobID: jobId, downloadPromise: promise })
    promise.then((r) => {
      console.log('Done downloading file', path, r.statusCode)
      this.setState({ jobID: null, filePath: r.statusCode === 200 ? `file://${path}` : null })
    })
  }

  handleLayout = (event: any) => {
    console.log('onLayout', event)
    const { width, height } = event.nativeEvent.layout
    if (height !== 0 && width !== this.state.size.width && height !== this.state.size.height) {
      this.setState({ size: { width, height } })
    }
  }

  componentWillUnmount () {
    if (this.state.jobID) {
      stopDownload(this.state.jobID)
    }
  }

  done = () => {
    this.props.navigator.dismiss()
  }

  share = () => {
    if (this.state.filePath) {
      ActionSheetIOS.showShareActionSheetWithOptions({ url: this.state.filePath }, (error: Error) => {
        console.log('Failed showing share sheet', error)
      }, (success: boolean, method: string) => {
        console.log('Successfully shared file', method)
      })
    }
  }

  renderBody () {
    let body = <View></View>
    switch (this.props.attachment.mime_class) {
      case 'image':
        body = <View style={styles.imageContainer}>
          <Image source={{ uri: this.state.filePath }} resizeMode='contain' style={styles.image} />
        </View>
        break
      case 'audio':
      case 'video':
        body = this.renderAudioVisual()
        break
      case 'zip':
      case 'flash':
        body = <View style={styles.centeredContainer}>
          <Text style={{ textAlign: 'center', color: Colors.darkText, fontSize: 14 }}>{i18n('Previewing this file type is not supported')}</Text>
        </View>
        break
      default:
        if (this.props.attachment['content-type'] && this.props.attachment['content-type'].indexOf('audio') !== -1) {
          body = this.renderAudioVisual()
        } else {
          body = <WebView source={{ uri: this.state.filePath }} style={styles.document} />
        }
    }
    return body
  }

  renderAudioVisual () {
    return (
      <View style={styles.centeredContainer} onLayout={this.handleLayout}>
        <Video
          source={{ uri: this.state.filePath }}
          style={{ width: this.state.size.width, height: Math.ceil(this.state.size.width * 9.0 / 16.0) }}
        />
      </View>
    )
  }

  render () {
    return (
      <Screen
        title={i18n('Attachment')}
        navBarStyle='light'
        navBarTitleColors={Colors.darkText}
        navBarButtonColor={Colors.link}
        drawUnderNavBar={true}
        leftBarButtons={[{
          testID: 'attachment-view.done-btn',
          title: i18n('Done'),
          style: 'done',
          action: this.done,
        }]}
        rightBarButtons={[{
          testID: 'attachment-view.share-btn',
          image: Images.share,
          action: this.share,
        }]}
      >
        <View style={styles.container}>
          { this.state.filePath === null &&
            <View style={styles.centeredContainer}>
              <ActivityIndicator />
            </View>
          }
          { this.state.filePath !== null &&
            this.renderBody()
          }
        </View>
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  centeredContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  imageContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    backgroundColor: Colors.grey1,
  },
  image: {
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    position: 'absolute',
  },
  document: {
    flex: 1,
  },
})
