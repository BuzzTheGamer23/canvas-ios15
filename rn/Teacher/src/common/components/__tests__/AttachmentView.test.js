//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import React from 'react'
import { ActionSheetIOS } from 'react-native'
import AttachmentView from '../AttachmentView'
import renderer from 'react-test-renderer'
import md5 from 'md5'

const templates = {
  ...require('../../../__templates__/helm'),
}

jest.mock('ActionSheetIOS', () => ({
  showShareActionSheetWithOptions: jest.fn(),
}))

jest.mock('react-native-fs', () => ({
  CachesDirectoryPath: 'caches',
  downloadFile: jest.fn(() => ({
    jobId: '1',
    promise: Promise.resolve({ statusCode: 200 }),
  })),
}))

jest.mock('WebView', () => 'WebView')

let defaultProps = {
  navigator: templates.navigator(),
  attachment: {
    id: '1',
    display_name: 'Something.pdf',
    url: 'http://www.fillmurray.com/100/100',
    mime_class: 'pdf',
  },
}

describe('AttachmentView', () => {
  it('renders a pdf', async () => {
    let tree = renderer.create(
      <AttachmentView {...defaultProps} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders an image', async () => {
    let props = {
      navigator: templates.navigator(),
      attachment: {
        id: '1',
        display_name: 'Something.png',
        url: 'http://www.fillmurray.com/100/100',
        mime_class: 'image',
      },
    }

    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders unsupported stuffs', async () => {
    let props = {
      navigator: templates.navigator(),
      attachment: {
        id: '1',
        display_name: 'Something.zip',
        url: '',
        mime_class: 'zip',
      },
    }

    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders a video', async () => {
    let props = {
      navigator: templates.navigator(),
      attachment: {
        id: '1',
        display_name: 'Something.mp4',
        url: '',
        mime_class: 'video',
      },
    }

    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders audio', async () => {
    let props = {
      navigator: templates.navigator(),
      attachment: {
        id: '1',
        display_name: 'Something.mp3',
        url: '',
        mime_class: 'audio',
      },
    }

    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders audio with unknown mime class', async () => {
    let props = {
      navigator: templates.navigator(),
      attachment: {
        id: '1',
        display_name: 'Something.mp3',
        url: '',
        mime_class: 'apple-audio',
        'content-type': 'audio/mp3',
      },
    }

    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    await tree.getInstance().fetchFile()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('opens share sheet', () => {
    const props = {
      ...defaultProps,
      navigator: templates.navigator({
        dismiss: jest.fn(),
      }),
    }
    let tree = renderer.create(
      <AttachmentView {...props} />
    )
    let instance = tree.getInstance()

    return instance.state.downloadPromise.then(r => {
      instance.share()
      expect(ActionSheetIOS.showShareActionSheetWithOptions).toHaveBeenCalledWith({
        url: `file://caches/${md5(props.attachment.url)}.pdf`,
      }, expect.any(Function), expect.any(Function))
    })
  })

  it('closes', () => {
    const props = {
      ...defaultProps,
      navigator: templates.navigator({
        dismiss: jest.fn(),
      }),
    }
    let tree = renderer.create(
      <AttachmentView {...props} />
    )

    tree.getInstance().done()
    expect(props.navigator.dismiss).toHaveBeenCalled()
  })
})
