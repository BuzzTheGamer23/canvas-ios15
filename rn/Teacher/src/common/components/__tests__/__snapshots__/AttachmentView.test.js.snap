// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentView renders a pdf 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <WebView
    source={
      Object {
        "uri": "file://caches/ed075ce5e86325eed0f5138a4613fdd6.pdf",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`AttachmentView renders a video 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <VideoContainer
      source={
        Object {
          "uri": "file://caches/d41d8cd98f00b204e9800998ecf8427e.mp4",
        }
      }
      style={
        Object {
          "height": 0,
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`AttachmentView renders an image 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5f5f5",
        "flex": 1,
        "justifyContent": "center",
        "position": "relative",
      }
    }
  >
    <Image
      onError={[Function]}
      resizeMode="contain"
      source={
        Object {
          "uri": "file://caches/ed075ce5e86325eed0f5138a4613fdd6.png",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`AttachmentView renders audio 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <VideoContainer
      source={
        Object {
          "uri": "file://caches/d41d8cd98f00b204e9800998ecf8427e.mp3",
        }
      }
      style={
        Object {
          "height": 0,
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`AttachmentView renders audio with unknown mime class 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <VideoContainer
      source={
        Object {
          "uri": "file://caches/d41d8cd98f00b204e9800998ecf8427e.mp3",
        }
      }
      style={
        Object {
          "height": 0,
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`AttachmentView renders unsupported stuffs 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#2D3B45",
          "fontSize": 14,
          "textAlign": "center",
        }
      }
    >
      Previewing this file type is not supported
    </Text>
  </View>
</View>
`;
